src/ripple/conditions/impl/utils.h

```diff
@@ -31,338 +33,195 @@
+    s += 1;
 
-    for (std::size_t i = 0; i < sizeof(Integer); ++i)
-        res = (res << 8) | *begin++;
+    if (p.tag == 0x1F)
+    { // Long tag form:
+        p.tag = 0;
 
-    return { begin, res };
-}
+        if (s[0] == 0x80)
+        {
+            ec = error::generic;
+            return p;
+        }
 
-template <class OutputIt>
-inline
-OutputIt
-encode_length (std::size_t len, OutputIt out)
-{
-    if (len <= 0x7F)
-    {
-        *out++ = static_cast<std::uint8_t>(len & 0x7F);
-        return out;
-    }
+        while (!s.empty() && (s[0] & 0x80))
+        {
+            p.tag = p.tag * 128 + (s[0] & 0x7F);
+            s += 1;
+        }
 
-    // Decide how many bytes we need:
-    if (len <= 0xFFFF)
-    {
-        *out++ = 0x82;
-        *out++ = static_cast<std::uint8_t>((len >> 8) & 0xFF);
-        *out++ = static_cast<std::uint8_t>(len & 0xFF);
-        return out;
-    }
+        if (s.empty())
+        {
+            ec = error::generic;
+            return p;
+        }
 
-    if (len <= 0xFFFFFF)
-    {
-        *out++ = 0x83;
-        *out++ = static_cast<std::uint8_t>((len >> 16) & 0xFF);
-        *out++ = static_cast<std::uint8_t>((len >> 8) & 0xFF);
-        *out++ = static_cast<std::uint8_t>(len & 0xFF);
-        return out;
+        p.tag = p.tag * 128 + s[0];
+        s += 1;
     }
 
-    if (len <= 0xFFFFFFFF)
+    if (s.empty())
     {
-        *out++ = 0x84;
-        *out++ = static_cast<std::uint8_t>((len >> 24) & 0xFF);
-        *out++ = static_cast<std::uint8_t>((len >> 16) & 0xFF);
-        *out++ = static_cast<std::uint8_t>((len >> 8) & 0xFF);
-        *out++ = static_cast<std::uint8_t>(len & 0xFF);
-        return out;
+        ec = error::generic;
+        return p;
     }
 
-    // Note: OER can represent lengths up to (2^1016) - 1,
-    // which is, truly, enough for everyone. We never
-    // exceed 2^32.
-    Throw<std::length_error>("overlong encoding length: " + std::to_string(len));
-}
+    p.length = s[0];
+    s += 1;
 
-// A "streambuf" would serve us better here - instead of the
-// crazy paired return, we consume data from it and things
-// just magically work.
-template <class InputIt>
-std::pair<InputIt, std::size_t>
-decode_length (InputIt begin, InputIt end)
-{
-    if (begin == end)
-        Throw<std::length_error>("empty buffer");
-
-    std::size_t bytes = *begin++;
+    if (p.length & 0x80)
+    { // Long form length:
+        std::size_t const cnt = p.length & 0x7F;
 
-    if (bytes < 128)
-        return { begin, bytes };
+        if (cnt == 0)
+        {
+            ec = error::generic;
+            return p;
+        }
 
-    bytes &= 0x7F;
+        if (cnt > sizeof(std::size_t))
+        {
+            ec = error::generic;
+            return p;
+        }
 
-    if (bytes > 4)
-        Throw<std::length_error>("overlong encoded length: " + std::to_string(bytes));
+        // Account for the first two bytes in the preamble:
+        if (cnt >= s.size())
```

Comment:
We might want to check the case where `cnt` is `sizeof(std::size_t)` and `s` is greater.